tracker/view/login.py

```diff
@@ -29,9 +35,36 @@ def login():
                                User=User,
                                password_length={'min': TRACKER_PASSWORD_LENGTH_MIN,
                                                 'max': TRACKER_PASSWORD_LENGTH_MAX}), status_code
-
+    if TRACKER_ENABLE_TWO_FACTOR:
+        session['two_factor'] = form.user.name
+        return redirect(url_for('tracker.two_factor'))
     user = user_assign_new_token(form.user)
-    user.is_authenticated = True
+    login_user(user)
+    return redirect(url_for('tracker.index'))
+
+
+@tracker.route('/twofactor', methods=['GET', 'POST'])
+def two_factor():
+    if current_user.is_authenticated:
+        return redirect(url_for('tracker.index'))
+    if not session.get('two_factor'):
+        return redirect(url_for('tracker.login'))
+    user = User.query.filter(User.name == session['two_factor']).first()
+    if not user.otp_token:
+        user.otp_token = pyotp.random_base32()
+        user.first_time_login_otp = True
+    form = OTPForm()
+    if not form.validate_on_submit():
+        status_code = Unauthorized.code if form.is_submitted() else 200
+        return render_template('two_factor.html',
+                               title='Two Factor Login',
+                               User=User,
+                               form=form,
+                               display_qr=current_user.first_time_login_otp,
```

Comment:
This isn't going to work anymore as `current_user` isn't a source for information unless we call `login_user(user)`.